marda_extractors_api/__init__.py

```diff
@@ -0,0 +1,155 @@
+import subprocess
+from enum import Enum
+from pathlib import Path
+
+import httpx
+
+__all__ = ("extract", "MardaExtractor")
+
+REGISTRY_BASE_URL = "https://marda-registry.fly.dev"
+
+
+class SupportedExecutionMethod(Enum):
+    CLI = "cli"
+    PYTHON = "python"
+
+
+def extract(
+    file_path: Path,
+    file_type: str,
+    output_file: Path | None = None,
+    preferred_mode: SupportedExecutionMethod = SupportedExecutionMethod.PYTHON,
+):
+    """Parse a file given its path and file type ID
+    in the MaRDA registry.
+    """
+
+    if not file_path.exists():
+        raise RuntimeError(f"File {file_path} does not exist")
+
+    response = httpx.get(f"{REGISTRY_BASE_URL}/filetypes/{file_type}")
+    if response.status_code != 200:
+        raise RuntimeError(
+            f"Could not find file type {file_type!r} in the registry at {response.url!r}"
+        )
+    extractors = response.json()["registered_extractors"]
+    if not extractors:
+        raise RuntimeError(
+            f"No extractors found for file type {file_type!r} in the registry"
+        )
+    elif len(extractors) > 1:
+        print(
+            f"Discovered multiple extractors: {extractors}, using the first ({extractors[0]})"
+        )
+
+    extractor = extractors[0]
+    entry = httpx.get(f"{REGISTRY_BASE_URL}/extractors/{extractor}")
+    if response.status_code != 200:
+        raise RuntimeError(f"Could not find extractor {extractor!r} in the registry")
+
+    extractor = MardaExtractor(entry.json(), preferred_mode=preferred_mode)
+
+    return extractor.execute(file_type, file_path, output_file)
+
+
+class MardaExtractor:
+    """A plan for parsing a file."""
+
+    entry: dict
+    """The registry entry to use for parsing."""
+
+    def __init__(
+        self,
+        entry: dict,
+        install: bool = True,
+        preferred_mode: SupportedExecutionMethod = SupportedExecutionMethod.PYTHON,
+    ):
+        """Initialize the plan, optionally installing the specific parser package."""
+        self.entry = entry
+        self.preferred_mode = preferred_mode
+        if install:
+            self.install()
+
+    def install(self):
+        command = ["pip", "install", f"git+{self.entry['source_repository']}"]
+        subprocess.check_output(command)
+
+    def execute(self, file_type: str, file_path: Path, output_file: Path | None = None):
+        if file_type not in [d["id"] for d in self.entry["supported_filetypes"]]:
+            raise ValueError(
+                f"File type {file_type!r} not supported by {self.entry['id']!r}"
+            )
+
+        if self.entry["id"] == "yadg":
+            self.entry["usage"].append(
+                "python:yadg.extractors.extract({{ file_type }}, {{ file_path }})"
+            )
```

Comment:
Let's get rid of this and add it to the yadg `Extractor`. I will also add `mpt` and `mpr` example files.